### PR TITLE
Implement pagination on the user list

### LINKED ISSUE #20: Error: Too Many Requests
I've been trying to fix our https://github.com/rauchg/slackin instance, and after spending a lot of time fixing simple but hard to spot issues, I keep getting this error. So, then I googled around and found your fork @emedvedev :)

I decided to give it a go and I'm still getting:

```
2019-10-19T15:50:35.973813+00:00 heroku[web.1]: Starting process with command `bin/slackin --coc "$SLACKIN_COC" --channels "" getbootstrap FOO`
2019-10-19T15:50:38.829158+00:00 app[web.1]: 2019-10-19T15:50:38.827Z slackin:main Theme: light
2019-10-19T15:50:38.837343+00:00 app[web.1]: 2019-10-19T15:50:38.837Z slackin:main Establishing connection with Slack
2019-10-19T15:50:38.869597+00:00 app[web.1]: 2019-10-19T15:50:38.869Z slackin:slack Fetching data from Slack
2019-10-19T15:50:38.892309+00:00 app[web.1]: 2019-10-19T15:50:38.892Z slackin:main Listening on 0.0.0.0:PORT
2019-10-19T15:50:38.953810+00:00 app[web.1]: 2019-10-19T15:50:38.953Z slackin:slack Error: Error: Too Many Requests
2019-10-19T15:50:38.953814+00:00 app[web.1]: at Request.callback (/app/node_modules/superagent/lib/node/index.js:706:15)2019-10-19T15:50:38.953816+00:00 app[web.1]: at parser (/app/node_modules/superagent/lib/node/index.js:916:18)
2019-10-19T15:50:38.953818+00:00 app[web.1]: at IncomingMessage.res.on (/app/node_modules/superagent/lib/node/parsers/json.js:19:7)
2019-10-19T15:50:38.953821+00:00 app[web.1]: at emitNone (events.js:111:20)
2019-10-19T15:50:38.953823+00:00 app[web.1]: at IncomingMessage.emit (events.js:208:7)
2019-10-19T15:50:38.953827+00:00 app[web.1]: at endReadableNT (_stream_readable.js:1064:12)
2019-10-19T15:50:38.953830+00:00 app[web.1]: at _combinedTickCallback (internal/process/next_tick.js:139:11)
2019-10-19T15:50:38.953832+00:00 app[web.1]: at process._tickCallback (internal/process/next_tick.js:181:9)
2019-10-19T15:50:38.954667+00:00 app[web.1]: 2019-10-19T15:50:38.954Z slackin:slack Attempt failed, will retry
2019-10-19T15:50:39.773455+00:00 heroku[web.1]: State changed from starting to up
2019-10-19T15:51:27.050728+00:00 heroku[router]: at=error code=H12 desc="Request timeout" method=GET path="/badge.svg" host=bootstrap-slack.herokuapp.com request_id=XXX fwd="XXX" dyno=web.1 connect=1ms service=30000ms status=503 bytes=0 protocol=https
```

Maybe the Slack API calls could be rate limited?

Thanks in advance!

PS. #19 is the other issue I hit while trying your fork.

### PR DESCRIPTION
**Pull request by emedvedev**
Fixes #20

@XhmikosR sorry for such a delay! But I'm finally ready to get back to this :)

This PR implements cursor-based pagination on the user list, and while it's not perfect yet, should already work. 

To do:
- [x] Retry a specific page on a rate limiting event, not the entire thing (which is probably useless).
- [x] Implement the same for the channel list (would anyone even need pagination for the channel list though?)
- [x] Handle `{"ok": false}` responses as errors.
- [x] Make delay between pages configurable

### CODE REVIEW

**Comment by emedvedev** [comment]
@XhmikosR can you try this out when you have a bit of time and tell me if it worked for your workspace? I don't have any large workspaces I could test this on, so I have to test it with very small page sizes (3–5 results per page) that wouldn't be rate-limited to begin with.

**Comment by XhmikosR** [comment]
Yup, definitely! Let's move with the other easy stuff and I will let you know about this

File: lib/slack.js [hunk 1]
```diff
@@ -49,43 +49,41 @@ class SlackData extends EventEmitter {
...
+        return this.retry();
+      }
+      if (!response.body.members) {
+        this.emit('error', new Error(`Invalid Slack response: ${response.status}`));
+        return this.retry();
+      }
+      users = users.concat(response.body.members);
+      cursor = response.body.response_metadata.next_cursor;
     }
+    while (firstPage === true || cursor !== '');
```
Comment (XhmikosR) [inline lib/slack.js]: Are you sure we need the `firstPage` variable? We should just loop until `cursor !== ''`.

Comment (emedvedev) [reply lib/slack.js]: Ah good catch. That was a remnant from a less elegant way I've tried first :)

**Comment by XhmikosR** [comment]
@emedvedev I tried it locally and it appears to be stuck since it tries to make the requests sequentially, but it does throw after a while:

```
> slackin-extended@1.1.11 start C:\Users\xmr\Desktop\slackin-extended
> cross-env-shell node ./bin/slackin $SLACK_SUBDOMAIN $SLACK_API_TOKEN --coc \"$SLACKIN_COC\" --channels \"$SLACKIN_CHANNELS\"

  slackin:main Theme: light +0ms
  slackin:main Establishing connection with Slack +6ms
  slackin:slack Fetching data from Slack +0ms
  slackin:main Listening on 0.0.0.0:3000 +0ms
  slackin:slack Error: Error: Too Many Requests
    at Request.callback (C:\Users\xmr\Desktop\slackin-extended\node_modules\superagent\lib\node\index.js:706:15)
    at C:\Users\xmr\Desktop\slackin-extended\node_modules\superagent\lib\node\index.js:916:18
    at IncomingMessage.<anonymous> (C:\Users\xmr\Desktop\slackin-extended\node_modules\superagent\lib\node\parsers\json.js:19:7)
    at IncomingMessage.emit (events.js:215:7)
    at endReadableNT (_stream_readable.js:1184:12)
    at processTicksAndRejections (internal/process/task_queues.js:80:21) +15s
  slackin:slack Attempt failed, will retry +1ms
```

**Comment by emedvedev** [comment]
Can you try adding a delay between requests (100-200ms?) and a log line to check how many requests it can make before it throws?

**Comment by XhmikosR** [comment]
Can you push the changes please?

**Comment by emedvedev** [comment]
Done :)

**Comment by emedvedev** [comment]
`users.list` should have Tier 2 rate limiting, which is "at least 20 requests per minute, while allowing for occasional bursts of more requests". Theoretically, a delay of 3000ms between page requests should always work.

Also, looks like the `response.status === 429` case doesn't work anymore, seeing as you're getting 429 but the requests don't retry properly. Should go into the `catch` statement, I'll try it in a bit.

**Comment by XhmikosR** [comment]
Now it just keeps running multiple times and is super slow :/ I'm not sure how we can to this faster without hitting the limits...

```
> slackin-extended@1.1.11 start C:\Users\xmr\Desktop\slackin-extended
> cross-env-shell node ./bin/slackin $SLACK_SUBDOMAIN $SLACK_API_TOKEN --coc \"$SLACKIN_COC\" --channels \"$SLACKIN_CHANNELS\"

  slackin:main Theme: light +0ms
  slackin:main Establishing connection with Slack +8ms
  slackin:slack Fetching data from Slack +0ms
  slackin:main Listening on 0.0.0.0:3000 +0ms
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +2s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +2s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +2s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +2s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Error: Error: Too Many Requests
    at Request.callback (C:\Users\xmr\Desktop\slackin-extended\node_modules\superagent\lib\node\index.js:706:15)
    at C:\Users\xmr\Desktop\slackin-extended\node_modules\superagent\lib\node\index.js:916:18
    at IncomingMessage.<anonymous> (C:\Users\xmr\Desktop\slackin-extended\node_modules\superagent\lib\node\parsers\json.js:19:7)
    at IncomingMessage.emit (events.js:215:7)
    at endReadableNT (_stream_readable.js:1184:12)
    at processTicksAndRejections (internal/process/task_queues.js:80:21) +1s
  slackin:slack Attempt failed, will retry +3ms
```

(sorry for the long post)

Maybe this helps? https://stackoverflow.com/a/55083898

**Comment by emedvedev** [comment]
How many users do you have? It should fetch up to 200 per request, otherwise it's probably an uncaught error.

I've pushed a worst-case delay of 3000ms between requests: it will be slow, but we should not hit rate limits this way. Can you retry?

**Comment by XhmikosR** [comment]
We have around 5K users. The problem is that this so slow that is unusable :/

Imagine (5000 / 200 ) * 4 seconds

**Comment by XhmikosR** [comment]
Correction we have ~30K people.

**Comment by emedvedev** [comment]
Ahh. I'm not sure what we can do in this case, since it's an API limitation. 

Still, it should be slow the first time around, and then it's just stored until the new numbers are fetched, so the first time will be a PITA, but after the launch it should be fine—the updates will be slow, but I don't think anyone would mind that the user count is not real-time, right?

**Comment by XhmikosR** [comment]
It still hasn't finished :P

**Comment by XhmikosR** [comment]
What I don't get is how this is working for them https://slack.rancher.io/

**Comment by XhmikosR** [comment]
OK now it did finish but it took a long time...

<details>
<summary>Log</summary>

```
> slackin-extended@1.1.11 start C:\Users\xmr\Desktop\slackin-extended
> cross-env-shell node ./bin/slackin $SLACK_SUBDOMAIN $SLACK_API_TOKEN --coc \"$SLACKIN_COC\" --channels \"$SLACKIN_CHANNELS\"

  slackin:main Theme: light +0ms
  slackin:main Establishing connection with Slack +6ms
  slackin:slack Fetching data from Slack +0ms
  slackin:main Listening on 0.0.0.0:3000 +0ms
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +5s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Slack is ready +753ms
  slackin:slack Got data from Slack: 0 online, 29756 total +1ms
  slackin:slack Fetching data from Slack +1m
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
  slackin:slack Fetching data from Slack +4s
```

</details>

**Comment by XhmikosR** [comment]
OK, this seems to work for me too, i.e. no errors and it's faster

```diff
 lib/slack.js | 5 +----
 1 file changed, 1 insertion(+), 4 deletions(-)

diff --git a/lib/slack.js b/lib/slack.js
index b984791..497a5a4 100644
--- a/lib/slack.js
+++ b/lib/slack.js
@@ -63,11 +63,8 @@ class SlackData extends EventEmitter {
         response = await request // eslint-disable-line no-await-in-loop
           .get(`https://${this.host}.slack.com/api/users.list`)
           .query({
-            token: this.token, limit: 200, cursor, presence: 1,
+            token: this.token, limit: 800, cursor, presence: 1,
           });
-        if (response.body.response_metadata && response.body.response_metadata.next_cursor !== '') {
-          await delay(3000); // eslint-disable-line no-await-in-loop
-        }
       } catch (err) {
         this.emit('error', err);
         if (err.response.status === 429) {

```

<details>
<summary>Log</summary>

```
> slackin-extended@1.1.11 start C:\Users\xmr\Desktop\slackin-extended
> cross-env-shell node ./bin/slackin $SLACK_SUBDOMAIN $SLACK_API_TOKEN --coc \"$SLACKIN_COC\" --channels \"$SLACKIN_CHANNELS\"

  slackin:main Theme: light +0ms
  slackin:main Establishing connection with Slack +7ms
  slackin:slack Fetching data from Slack +0ms
  slackin:main Listening on 0.0.0.0:3000 +0ms
  slackin:slack Fetching data from Slack +723ms
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +845ms
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +690ms
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +713ms
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +692ms
  slackin:slack Fetching data from Slack +705ms
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +722ms
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +706ms
  slackin:slack Fetching data from Slack +705ms
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +1s
  slackin:slack Fetching data from Slack +719ms
  slackin:slack Slack is ready +1s
  slackin:slack Got data from Slack: 0 online, 29756 total +2ms
```
</summary>

**Comment by emedvedev** [comment]
Interesting! Thanks for the testing :) 

We can keep it at 800 users per page and no delay then, but I'll comment it out instead of removing just in case—maybe we'll have to get it back at some point.

**Comment by XhmikosR** [comment]
Alright, I tested my changes above a bit more and indeed after a while we do hit the API rate limit. So we do need the delay...

**Comment by XhmikosR** [comment]
I still don't get how other channels do it with so many users, like https://slack.rancher.io/

**Comment by emedvedev** [comment]
Ok so I think this is ready. Could you help with testing one more time? :)

The page size defaults to 800, and the delay is configured as `SLACKIN_PAGE_DELAY` in the env variables. Should also respect the `retry-after` header in the HTTP 429 responses. So setting `SLACKIN_PAGE_DELAY` to 3000 (worst case) should definitely work, but something smaller might work as well. The limit as described in the API docs is 20 requests per minute with "occasional bursts."

**Comment by XhmikosR** [comment]
Thanks for this, I'm gonna try it later and let you know. In the mean time can you merge the other PRs and then merge master to this branch?

BTW while testing this branch there are still quite big slowdowns. I'm gonna provide a screenshot later to see the graph from Heroku.

**Comment by emedvedev** [comment]
Sure thing! Just review the ones marked as draft, I think at least some of them can be merged.

**Comment by XhmikosR** [comment]
I deployed this and used `SLACKIN_INTERVAL: 120000` and `SLACKIN_PAGE_DELAY: 2000` and I'll see how it'll go.

BTW this isn't new but it seems we don't retrieve the online members properly:

```
2019-11-26T13:23:28.040760+00:00 app[web.1]: 2019-11-26T13:23:28.040Z slackin:slack Got data from Slack: 0 online, 29812 total
```

Here's a screenshot from Heroku so that you see the 30s spikes

Screenshot_2019-11-26 bootstrap-slack · Metrics Heroku

**Comment by XhmikosR** [comment]
I also switched to Node.js 12.x just now (I had to change the package.json engines version, not sure if I can do it another way without having to keep a custom patch around).

**Comment by emedvedev** [comment]
Ok, I know what the 30s spikes are about, I'll fix that in a bit.

File: readme.md [hunk 2]
```diff
@@ -86,6 +86,7 @@ Every CLI parameter, including mandatory arguments (workspace ID and token), can
 | --accent | -A | `SLACKIN_ACCENT` | `#e01563` for `light`, `#9a0e44` for `dark` | Accent color to use instead of a theme default |
 | --coc | -C | `SLACKIN_COC` | `''` | Full URL to a CoC that needs to be agreed to |
 | --css | -S | `SLACKIN_CSS` | `''` | Full URL to a custom CSS file to use on the main page |
+| | | `SLACKIN_PAGE_DELAY` | `0` | Delay between Slack API requests (may be required for large workspaces that hit the rate limit) |
```
Comment (XhmikosR) [inline readme.md]: Maybe mention the unit (ms)?

**Comment by XhmikosR** [comment]
@emedvedev friendly ping :) Still using this branch and I do notice the spikes, but all in all it does seem to work https://bootstrap-slack.herokuapp.com/

**Comment by emedvedev** [comment]
Alright, back to work. :) So fixing the spikes is a bit harder than expected, so in the meantime I'll merge this PR, then work on the spikes in a separate branch,.

**Comment by emedvedev** [comment]
About the online members: I don't think we can anymore, because Slack has removed the presence responses from user lists. Querying thousands of users individually isn't exactly ideal :)

File: readme.md (hunk 2 above)
Comment (emedvedev) [reply readme.md]: Good catch :)

**Comment by XhmikosR** [comment]
We could probably remove `presence: 1` maybe? And also does this mean this is moot now?

```js
const active = users.filter((user) => user.presence === 'active').length;
```

In our case this patch does work but takes around 1 min to get the first response using `SLACKIN_PAGE_DELAY: 2000`. I wish this was faster, but I can't think of any way.

Only the spikes are left :)

**Comment by emedvedev** [comment]
It is most likely moot, but the API docs are unclear on that, and it's still mentioned, so I'll leave it for now. I can imagine a bunch of cases where the user count is needed, just in case the Slack devs change their mind eventually :)

**Pull request merged**